xscreensaver: Add service module and suid wrapper patch

vancluever opened the pull request:
## Description of changes

This adds the NixOS module (from @aidalgol) and a patch for XScreenSaver that ensures that the suid wrapper for xscreensaver-auth is run correctly.

The patch is a simple update to drivers/xscreensaver.c that inserts /run/wrappers/bin before the `DEFAULT_PATH_PREFIX`, which is the directory for xscreensaver hacks/demos, and should be preserved.

The wrapper directory can be modified in the derivation, or even disabled.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Fixes #130218.

vancluever commented:
Fixes #130218

vancluever commented:
@aidalgol I took your NixOS module and tests and expanded upon them :)

You're on here as co-contributor but let me know if you want it amended! I switched maintainer to myself though.

vancluever commented:
Also re: release notes, I haven't added anything since 23.11 is pretty close and there's no notes for 24.05 yet. But happy to add anything if need be!

vancluever commented:
Result of  `nix run nixpkgs#nixpkgs-review -- rev HEAD`:

1 package blacklisted:
nixos-install-tools

5 packages built:
caffeine-ng deepin.deepin-screensaver lumina.lumina xscreensaver xsecurelock

vancluever commented:
Tested with local copy of module here and looks like everything is good.

AndersonTorres commented on nixos/tests/xscreensaver.nix:
```diff
@@ -0,0 +1,64 @@
+import ./make-test-python.nix ({ pkgs, lib, ... }: {
```
Tests should be in a separated commit

vancluever replied:
Done! Also split out the package and the module while I was at it.

AndersonTorres commented on nixos/modules/services/x11/xscreensaver.nix:
```diff
@@ -0,0 +1,42 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
```suggestion
```

Do not use global with.

AndersonTorres commented on pkgs/by-name/xs/xscreensaver/package.nix:
```diff
@@ -26,6 +26,11 @@
+
+  # This is only included in the rare case it needs to be tuned, or if you want
+  # to exclude it altogether (just set to empty string)
+, withWrapperPrefix ? null
```
Do not use `null`.

AndersonTorres commented on pkgs/by-name/xs/xscreensaver/package.nix:
```diff
@@ -67,10 +72,17 @@ stdenv.mkDerivation (finalAttrs: {
     sed -e 's%@GTK_DATADIR@%@datadir@% ; s%@PO_DATADIR@%@datadir@%' \
       -i driver/Makefile.in po/Makefile.in.in
+  '' + lib.optionalString (builtins.isString withWrapperPrefix) ''
+    sed -E 's%^(#define NIXOS_WRAPPER_PREFIX).*$%\1 "${withWrapperPrefix}"%' \
```
You should use  `substituteAll` function instead.

7c6f434c commented on pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch:
```diff
@@ -0,0 +1,32 @@
+   char *npath;
+   if (! opath) opath = "/bin:/usr/bin";  /* WTF */
+-  npath = (char *) malloc(strlen(def_path) + strlen(opath) + 20);
++  npath = (char *) malloc(strlen(wrapper_path) + strlen(def_path) + strlen(opath) + 24);
```
Now I can't help wondering if 20 is _less_ than intended: something with UTF-32 default would need 28 bytes just for «PATH=», «:», and final `\0`. Of course, we all hope this doesn't exist and it was just UTF-16 + a bit of margin…

vancluever replied:
Yeah I guess it's never come up as it's in a lot of places in the code (e.g. margins of varying lengths). Given that, someone probably would need to ping jwz about it if it was a major concern :slightly_smiling_face:

AndersonTorres replied:
I would add 32 and a comment like "defensive extra room".

vancluever replied:
Done now!

vancluever commented on pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch:
```diff
@@ -0,0 +1,32 @@
++  npath = (char *) malloc(strlen(wrapper_path) + strlen(def_path) + strlen(opath) + 24);
+   strcpy (npath, "PATH=");
++  if (wrapper_path && *wrapper_path)
++    {
```
Yep; you can also see the same style in `hack_environment` in [`driver/demo-Gtk.c`](https://github.com/Zygo/xscreensaver/blob/master/driver/demo-Gtk.c#L3711-L3726) which is where this code was derived.

(Note that I didn't patch `demo-Gtk.c` as it seems that it's only used for demos and not `xscreensaver` or `xscreensaver-auth`).

vancluever commented:
Reviewing the build error and making correction.

vancluever commented:
**Updated** Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>caffeine-ng</li>
    <li>deepin.deepin-screensaver</li>
    <li>lumina.lumina</li>
    <li>xscreensaver</li>
    <li>xscreensaver.man</li>
    <li>xsecurelock</li>
  </ul>
</details>

vancluever commented:
@AndersonTorres just following up - have you been able to look at this again? Let me know if anything else needs done.

AndersonTorres reviewed (approved):
lgtm

vancluever commented:
:tada: Thanks @AndersonTorres!

AndersonTorres commented:
As the latest task,  can you rebase this against master? You were working with Xscreensaver 6.06 whilst the current is 6.08:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/xs/xscreensaver/package.nix

vancluever commented:
@AndersonTorres done now!

AndersonTorres commented:
Now squash the commits according to the contributing guidelines.

vancluever committed:
maintainers: add vancluever

vancluever committed:
xscreensaver: add suid wrapper patch

This adds a patch for XScreenSaver that ensures that the suid wrapper
for xscreensaver-auth is run correctly.

The patch is a simple update to drivers/xscreensaver.c that inserts
/run/wrappers/bin before the DEFAULT_PATH_PREFIX, which is the directory
for xscreensaver hacks/demos, and should be preserved.

The wrapper directory can be modified in the derivation, or even
disabled.

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

vancluever committed:
nixos/xscreensaver: init module

This adds a NixOS module for XScreenSaver (from @aidalgol in #130218,
with a few updates).

The module:

* Installs XScreenSaver
* Sets up a suid wrapper for xscreensaver-auth
* Sets up a user service for xscreensaver

The suid wrapper should function correctly when xscreensaver is
installed via the derivation update in 40a00547b71.

Co-authored-by: Aidan Gauland <aidalgol@fastmail.net>
Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

vancluever committed:
nixos/xscreensaver: add module tests

Co-authored-by: Aidan Gauland <aidalgol@fastmail.net>
Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

vancluever commented:
Thanks for the support in this @AndersonTorres and @7c6f434c!
